### PR TITLE
fix(hogql): derive cache keys from pydantic model_dump

### PR DESCRIPTION
**Pull request by thmsobrmlr**
## Problem

The PR https://github.com/PostHog/posthog/pull/22419 introduced a `clean_query` method to generate semantically equal queries for deriving a cache key. Since "clean..." methods are brittle, we're moving to a different approach here and use Pydantic's model serializer to strip out default values instead. 

This works for 80% of discrepancies and we can add in `@model_serializer`s to handle the remaining 20%. For example I'd like to follow up with a serializer that strips out [frontend-only options](https://github.com/PostHog/posthog/blob/b547d4ad59d9445002f4eacbc8f46b426be9f7ff/frontend/src/scenes/insights/utils/compareInsightQuery.ts#L46-L75) that don't affect the query output.

## Changes

Replaces the `clean_query` method with default values in the Pydantic models.

A couple of things were necessary to get this working:
- Our schema is generated with `Literal` fields for type, kind, etc. These are stripped by the `exclude_defaults=True` option, so we need to add them back in with a `@model_serializer`.
- Pydantic does not reliably sort keys with its `model_dump_json` method, so we need to use `orjson` with the respective setting.
- Typescript does not distinguish between float and integer types. As such when setting a default value for an integer type, it needs to be marked as such (`@asType integer` or using the `integer` type alias), as it will otherwise get translated into a float otherwise and this might break things.
- We needed to enable the `--set-default-enum-member` option for datamodel-codegen, to translate defaults for enums into enum values and not strings.

:warning: This will invalidate all insight cache keys again.

## How did you test this code?

Adapted tests and clicked around

### CODE REVIEW

**Review by aspicer — APPROVED** [review]
I like the approach of mostly correct cache keys to avoid the potential errors that come with the cleaner

File: posthog/schema_helpers.py [hunk 1]
```diff
@@ -1,81 +1,14 @@
...
-    FunnelsQuery,
-    FunnelVizType,
-    FunnelConversionWindowTimeUnit,
-    FunnelLayout,
-    BreakdownAttributionType,
-    StepOrderValue,
-)
 
+def to_json(query: BaseModel) -> bytes:
+    instance_dict = query.model_construct(**query.model_dump(exclude_none=True, exclude_defaults=False)).model_dump(
```
Comment (webjunkie) [inline posthog/schema_helpers.py:7]: Sorry, not a helpful comment, but this looks like a lot of dumping around just to get to a cache key in the end 🙈

Comment (thmsobrmlr) [reply posthog/schema_helpers.py]: Sorry, that's already outdated. See new code for a separate approach (manually re-adding the literal defaults). I'm open for other suggestions though - haven't figured out a clean solution that doesn't require us to write the models backend side yet.

File: bin/build-schema-python.sh [hunk 2]
```diff
@@ -7,7 +7,9 @@ datamodel-codegen \
     --class-name='SchemaRoot' --collapse-root-models --target-python-version 3.10 --disable-timestamp \
     --use-one-literal-as-default --use-default --use-default-kwarg --use-subclass-enum \
     --input frontend/src/queries/schema.json --input-file-type jsonschema \
-    --output posthog/schema.py --output-model-type pydantic_v2.BaseModel
+    --output posthog/schema.py --output-model-type pydantic_v2.BaseModel \
+    --custom-file-header "# mypy: disable-error-code=\"assignment\"" \
+    --set-default-enum-member
```
Comment (thmsobrmlr) [inline bin/build-schema-python.sh:12]: This is necessary to translate a default like

```js
/** @default numeric */
aggregationAxisFormat?: AggregationAxisFormat
```

into an enum like

```py
aggregationAxisFormat: Optional[AggregationAxisFormat] = AggregationAxisFormat.numeric
```

**Comment by thmsobrmlr** [comment]
Requesting another review here @PostHog/team-product-analytics, as the contents of the PR substantially changed since the last review.

File: frontend/src/queries/schema.json [hunk 3]
```diff
@@ -3752,14 +3752,14 @@
                     "default": {
                         "breakdown_type": "event"
                     },
-                    "description": "Breakdown of the events and actions"
+                    "description": "Breakdown of the events and actions\n\n:TRICKY: We can't set the default to an empty dict, as otherwise datamodel-code-generator does not generate a model with a factory.\n\nThis would cause a different serialization of an empty and a missing prop, which we don't want for stable cache keys.\n\nAs a workaround we simply set any prop to its default, so that a factory is used to instantiate an empty model for a missing prop."
```
Comment (webjunkie) [inline frontend/src/queries/schema.json:3755]: All these `TRICKY` comments will make it to the public OpenAPI schema as well... maybe there's a way to keep them out.

**Review by webjunkie — APPROVED** [review]
Looks alright to me. Just one thing.

File: frontend/src/queries/schema.json (hunk 3 above)
Comment (thmsobrmlr) [reply frontend/src/queries/schema.json]: Good point. I moved them to `// commments`, which removes them from the schema.

**Pull request merged**